File: solution/geeksforgeeks/NegativeWeightCycle.py
# https://practice.geeksforgeeks.org/problems/negative-weight-cycle/0

from math import inf

vertices, edges = 0,0
graph = [[] for _ in range(101)]
dist = [inf for _ in range(101)]

def initialise():
	global graph, dist
	for index, val in enumerate(graph):
		graph[index] = []
	for index, val in enumerate(dist):
		dist[index] = inf

def bellman_ford():
	global vertices, edges, graph, dist
	ans = False
	for index in range(vertices):
		for neighbour in graph[index]:
			if dist[neighbour[1]] > dist[index] + neighbour[0]:
				ans = True
				dist[neighbour[1]] = dist[index] + neighbour[0]
	return ans

def solve():
	global vertices, edges, graph
	vertices, edges = map(int, input().split())
	details = list(map(int, input().split()))
	dist[0] = 0
	for index in range(edges):
		source, dest, weight = details[index*3:index*3+3]
		graph[source].append((weight, dest))	
		# print(source, dest, weight)
	for _ in range(1,vertices):
		bellman_ford()
	print(1 if bellman_ford() is True else 0)
	# print(dist)
	# print(graph)

if __name__=='__main__':
	test = int(input())
	for tests in range(test):
		initialise()
		solve() 

File: solution/geeksforgeeks/NegativeWeightCycle.cpp
// https://practice.geeksforgeeks.org/problems/negative-weight-cycle/0

#include<bits/stdc++.h>
using namespace std;

#define ll long long int
#define pi pair<int,int>

#define inf 0x3f3f3f3f

int vertices, edges;
vector<pi> graph[101];
vector<int> dist(101,inf);

bool bellman_ford(){
    bool ans = false;
    for(int i=1;i<=vertices;i++){
        for(auto it:graph[i]){
            int neighbour = it.second, weight = it.first;
            if(dist[neighbour] > dist[i] + weight){
                ans = true;
                dist[neighbour] = dist[i] + weight;
            }
        }
    }
    return ans;
}

void init(){
    fill(dist.begin(),dist.end(),inf);
    for(int i=0;i<101;i++)
        graph[i].resize(0);
}

void solve(){
    cin >> vertices >> edges;
    for(int i=0;i<edges;i++){
        int from,to,weight;
        cin >> from >> to >> weight;
        graph[from+1].push_back(make_pair(weight, to+1));
    }
    for(int i=1;i<vertices;i++)
        bellman_ford();
    cout << bellman_ford() << '\n';
}

int main(int argc, char**argv){
    int test;
    cin >> test;
    while(test--){
        init();
        solve();
    }

    return 0;
}